src/vsr.zig

```diff
@@ -1575,6 +1575,28 @@ pub const Checkpoint = struct {
         assert(constants.journal_slot_count > constants.lsm_batch_multiple);
         assert(constants.journal_slot_count % constants.lsm_batch_multiple == 0);
     }
+    pub const Quorum = struct {
+        /// The latest known checkpoint operation from every *other* replica.
+        checkpoints: [constants.replicas_max]u128 = [_]u128{0} ** constants.replicas_max,
+
+        pub fn replace(quorum: *Quorum, replica: u8, checkpoint_op: u128) bool {
+            const checkpoint_op_existing: u128 = quorum.checkpoints[replica];
+            // Ignore old and repeat checkpoints.
+            if (checkpoint_op < checkpoint_op_existing or checkpoint_op == checkpoint_op_existing) {
+                return false;
+            }
+            quorum.checkpoints[replica] = checkpoint_op;
+            return true;
+        }
+
+        pub fn count(quorum: *const Quorum, checkpoint_op: u128) u8 {
```

Comment:
Something's wrong with types here. checkpoint_id is u128, but checkpoint_op is u64. Strincty speaking, we only need to track the op, but I'd track both op and id, if possible, just to have tighter verification.